File: python/examples/isolated_handler.py
import asyncio
import time
from collections.abc import AsyncGenerator
from datetime import datetime
from typing import Any

import uvloop

from camera_ui_rpc import create_rpc_client


async def main():
    print("Isolated Handler Connection Test\n")

    server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "main-server",
        }
    )

    await server.connect()
    print("Main server connected")

    async def echo(msg: str) -> str:
        print(f"[Handler] Echo received: {msg}")
        return f"Echo: {msg}"

    async def get_stats() -> dict[str, Any]:
        print("[Handler] Stats requested")
        return {
            "timestamp": datetime.now().isoformat(),
            "connections": 42,
            "uptime": 123.45,
        }

    async def slow_operation(delay: int) -> str:
        print(f"[Handler] Starting slow operation ({delay}ms)")
        await asyncio.sleep(delay / 1000)
        print("[Handler] Slow operation completed")
        return f"Completed after {delay}ms"

    async def generate_data(count: int) -> AsyncGenerator[dict[str, Any], None]:
        print(f"[Handler] Streaming {count} items")
        for i in range(count):
            yield {"index": i, "data": f"Item {i}"}
        print("[Handler] Stream completed")

    handlers: dict[str, Any] = {
        "echo": echo,
        "getStats": get_stats,
        "slowOperation": slow_operation,
        "generateData": generate_data,
    }

    print("\n--- Registering handlers with isolated connection ---")
    unsubscribe = await server.register_handler("isolated-test", handlers, isolated_connection=True)
    print("Handlers registered with isolated connection\n")

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-client",
        }
    )

    await client.connect()
    print("Client connected\n")

    proxy = client.create_proxy("isolated-test")

    try:
        total_start = time.perf_counter()

        print("--- Test 1: Echo ---")
        start = time.perf_counter()
        echo_result = await proxy.echo("Hello Isolated World!")
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Result: {echo_result} - Time: {elapsed:.1f}ms")

        print("\n--- Test 2: Get Stats ---")
        start = time.perf_counter()
        stats = await proxy.getStats()
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Stats: {stats} - Time: {elapsed:.1f}ms")

        print("\n--- Test 3: Concurrent Calls ---")
        start = time.perf_counter()
        tasks = [
            proxy.slowOperation(100),
            proxy.slowOperation(200),
            proxy.slowOperation(150),
            proxy.echo("Concurrent test"),
            proxy.getStats(),
        ]

        results = await asyncio.gather(*tasks)
        elapsed = (time.perf_counter() - start) * 1000
        print(f"All concurrent calls completed in {elapsed:.1f}ms:")
        for i, result in enumerate(results):
            if isinstance(result, dict):
                print(f"  {i + 1}: {result}")
            else:
                print(f"  {i + 1}: {result}")

        print("\n--- Test 4: Streaming ---")
        start = time.perf_counter()
        count = 0
        async for item in proxy.generateData(5):
            print(f"  Received: {item}")
            count += 1
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Streamed {count} items in {elapsed:.1f}ms")

        print("\n--- Test 5: Main Server Disconnect ---")
        print("Disconnecting main server...")
        await server.disconnect()
        print("Main server disconnected")

        print("Testing handlers after main disconnect...")
        try:
            await proxy.echo("Should fail")
            print("ERROR: Call should have failed")
        except Exception as e:
            print(f"Call failed as expected: {e}")

        print("\n--- Test 6: Unsubscribe Handlers ---")
        await unsubscribe()
        print("Handlers unsubscribed")

        try:
            await proxy.echo("Should fail")
            print("ERROR: Call should have failed")
        except Exception as e:
            print(f"Call failed as expected: {e}")

        total_elapsed = (time.perf_counter() - total_start) * 1000
        print(f"\nAll tests completed! Total time: {total_elapsed:.1f}ms")

    except Exception as e:
        print(f"Test failed: {e}")
        import traceback

        traceback.print_exc()
    finally:
        await client.disconnect()
        if server.is_connected:
            await server.disconnect()


if __name__ == "__main__":
    try:
        uvloop.run(main())
    except KeyboardInterrupt:
        print("\nTest interrupted by user")
    except Exception as e:
        print(f"Error: {e}")
        import traceback

        traceback.print_exc()

File: node/examples/isolated-handler.ts
import { createRPCClient } from '../src/index.js';

async function main() {
  console.log('=== Isolated Handler Connection Test ===\n');

  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'main-server',
  });

  await server.connect();
  console.log('Main server connected');

  const handlers = {
    echo: async (msg: string) => {
      console.log(`Echo received: ${msg}`);
      return `Echo: ${msg}`;
    },

    getStats: async () => {
      console.log('Stats requested');
      return {
        timestamp: new Date().toISOString(),
        connections: 42,
        uptime: process.uptime(),
      };
    },

    slowOperation: async (delay: number) => {
      console.log(`Starting slow operation (${delay}ms)`);
      await new Promise((resolve) => setTimeout(resolve, delay));
      console.log('Slow operation completed');
      return `Completed after ${delay}ms`;
    },

    async *generateData(count: number) {
      console.log(`Streaming ${count} items`);
      for (let i = 0; i < count; i++) {
        yield { index: i, data: `Item ${i}` };
      }
      console.log('Stream completed');
    },
  };

  const unsubscribe = await server.registerHandler('isolated-test', handlers, { isolatedConnection: true });
  console.log('Handlers registered with isolated connection\n');

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'test-client',
  });

  await client.connect();
  console.log('Client connected\n');

  const proxy = client.createProxy<{
    echo(msg: string): Promise<string>;
    getStats(): Promise<{ timestamp: string; connections: number; uptime: number }>;
    slowOperation(delay: number): Promise<string>;
    generateData(count: number): AsyncGenerator<{ index: number; data: string }>;
  }>('isolated-test');

  try {
    const totalStart = performance.now();

    console.log('Test 1: Echo');
    let start = performance.now();
    const echoResult = await proxy.echo('Hello Isolated World!');
    let elapsed = performance.now() - start;
    console.log(`Result: ${echoResult} - Time: ${elapsed.toFixed(1)}ms`);

    console.log('\nTest 2: Get Stats');
    start = performance.now();
    const stats = await proxy.getStats();
    elapsed = performance.now() - start;
    console.log(`Stats: ${JSON.stringify(stats)} - Time: ${elapsed.toFixed(1)}ms`);

    console.log('\nTest 3: Concurrent Calls');
    start = performance.now();
    const promises = [proxy.slowOperation(100), proxy.slowOperation(200), proxy.slowOperation(150), proxy.echo('Concurrent test'), proxy.getStats()];

    const results = await Promise.all(promises);
    elapsed = performance.now() - start;
    console.log(`All concurrent calls completed in ${elapsed.toFixed(1)}ms:`);
    results.forEach((result, i) => console.log(`  ${i + 1}: ${typeof result === 'object' ? JSON.stringify(result) : result}`));

    console.log('\nTest 4: Streaming');
    start = performance.now();
    const stream = proxy.generateData(5);
    let count = 0;
    for await (const item of stream) {
      console.log(`  Received: ${JSON.stringify(item)}`);
      count++;
    }
    elapsed = performance.now() - start;
    console.log(`Streamed ${count} items in ${elapsed.toFixed(1)}ms`);

    // Isolated handlers should also stop when the main server disconnects
    console.log('\nTest 5: Main Server Disconnect');
    console.log('Disconnecting main server...');
    await server.disconnect();
    console.log('Main server disconnected');

    console.log('Testing handlers after main disconnect...');
    try {
      await proxy.echo('Should fail');
      console.log('ERROR: Call should have failed');
    } catch (error) {
      console.log('Call failed as expected:', error instanceof Error ? error.message : error);
    }

    console.log('\nTest 6: Unsubscribe Handlers');
    await unsubscribe();
    console.log('Handlers unsubscribed');

    try {
      await proxy.echo('Should fail');
      console.log('ERROR: Call should have failed');
    } catch (error) {
      console.log('Call failed as expected:', error instanceof Error ? error.message : error);
    }

    const totalElapsed = performance.now() - totalStart;
    console.log(`\nAll tests completed! Total time: ${totalElapsed.toFixed(1)}ms`);
  } catch (error) {
    console.error('Test failed:', error);
  } finally {
    await client.disconnect();
    if (server.isConnected) {
      await server.disconnect();
    }
  }
}

main().catch(console.error);
